deflate.c

```diff
@@ -1144,9 +1140,9 @@ static void lm_set_level(deflate_state *s, int level) {
         s->insert_string = &insert_string_roll;
         s->quick_insert_string = &quick_insert_string_roll;
     } else {
-        s->update_hash = functable.update_hash;
-        s->insert_string = functable.insert_string;
-        s->quick_insert_string = functable.quick_insert_string;
+        s->update_hash = update_hash;
```

Comment:
Would it be possible to replace these function pointers as well?

Reply:
No, not without making separate functions for levels 7+8, and 9. Only level 9 uses the _roll functions due to doing 3-byte matching instead of the much faster 4-byte matching.

Reply:
ie: splitting deflate_slow() into two.

Reply:
Would just splitting longest_match_slow be worthwhile as a compromise between speed and complexity?

Reply:
Possibly. It would have to be benchmarked.

If I had to guess, I would say something around 3-5% faster for levels 7-9, up from their current ~1.1% speedup with this PR. I'd say that would be a good speedup, and might be worth it depending on how much more messy the code gets. I do not expect a significant increase in library size, so that'd probably be ok.

Reply:
@Dead2 

I think it would be better to split the implementations of ``deflate_quick``, ``deflate_medium``, ``deflate_slow`` into several functions and select it at runtime based on compression level and available processor features.

This will eliminate calls to the functions ``update_hash``, ``insert_string``, ``quick_insert_string`` (make it ``static inline``) by pointers.

I'll try to make prototype the implementation tomorrow. I think it will replace PR: #1681, #1682, #1686.

I'm guessing that hashing using crc CPU instructions might give better performance.

Reply:
Separating all functions greatly bloats the code. :(

It only makes sense to separate ``deflate_slow`` and ``PREFIX(fill_window)``. This will allow to inline the calls ``s->update_hash``, ``s->insert_string``, ``s->quick_insert_string``. We can call the splitted ``deflate_slow`` and ``fill_window`` through the pointers in ``deflate_state``.

This optimization can be implemented in this PR or on top of this PR.

Reply:
How much does it increase code size by?

Reply:
Call ``update_hash`` by pointer are present only in ``deflate_slow``.

I tried to implement my idea, but it did not increase performance. And the complexity of the code has increased significantly.